File: app/src/main/java/com/dingaimin/dedaocompanion/PlaybackCompletionPolicy.java
package com.dingaimin.dedaocompanion;

final class PlaybackCompletionPolicy {
    private static final long END_WINDOW_MS = 3_000L;

    static boolean reachedEnd(long furthestPosition, long duration) {
        return duration > 0L
                && Math.max(0L, furthestPosition) >= Math.max(0L, duration - END_WINDOW_MS);
    }

    static long probeDelay(long position, long duration) {
        return duration > 0L && position >= duration - 5_000L ? 250L : 1_000L;
    }

    static long updateFurthest(long previous, long position, boolean playing) {
        // A backward seek starts a fresh end window. A terminal zero, however, may be EOF.
        if (playing && position + END_WINDOW_MS < previous) return Math.max(0L, position);
        return Math.max(previous, Math.max(0L, position));
    }

    private PlaybackCompletionPolicy() {
    }
}

File: app/src/main/java/com/dingaimin/dedaocompanion/PlaybackCompletionPolicy.kt
package com.dingaimin.dedaocompanion

object PlaybackCompletionPolicy {
    private const val END_WINDOW_MS = 3_000L

    fun reachedEnd(furthestPosition: Long, duration: Long): Boolean =
        duration > 0 &&
            furthestPosition.coerceAtLeast(0) >= (duration - END_WINDOW_MS).coerceAtLeast(0)

    fun probeDelay(position: Long, duration: Long): Long =
        if (duration > 0 && position >= duration - 5_000L) 250L else 1_000L

    fun updateFurthest(previous: Long, position: Long, playing: Boolean): Long =
        if (playing && position + END_WINDOW_MS < previous) position.coerceAtLeast(0)
        else maxOf(previous, position, 0L)
}
